lnrpc/invoicesrpc/addinvoice.go

```diff
@@ -366,93 +372,35 @@ func AddInvoice(ctx context.Context, cfg *AddInvoiceConfig,
 
 	// We make sure that the given invoice routing hints number is within
 	// the valid range
-	if len(invoice.RouteHints) > 20 {
-		return nil, nil, fmt.Errorf("number of routing hints must " +
-			"not exceed maximum of 20")
-	}
-
-	// We continue by populating the requested routing hints indexing their
-	// corresponding channels so we won't duplicate them.
-	forcedHints := make(map[uint64]struct{})
-	for _, h := range invoice.RouteHints {
-		if len(h) == 0 {
-			return nil, nil, fmt.Errorf("number of hop hint " +
-				"within a route must be positive")
-		}
-		options = append(options, zpay32.RouteHint(h))
-
-		// Only this first hop is our direct channel.
-		forcedHints[h[0].ChannelID] = struct{}{}
+	if len(invoice.RouteHints) > maxHopHints {
+		return nil, nil, fmt.Errorf("number of routing hints must "+
+			"not exceed maximum of %v", maxHopHints)
 	}
 
-	// If we were requested to include routing hints in the invoice, then
-	// we'll fetch all of our available private channels and create routing
-	// hints for them.
-	if invoice.Private {
-		openChannels, err := cfg.ChanDB.FetchAllChannels()
-		if err != nil {
-			return nil, nil, fmt.Errorf("could not fetch all " +
-				"channels")
+	// Include route hints if needed.
+	if len(invoice.RouteHints) > 0 || invoice.Private {
+		totalHopHints := len(invoice.RouteHints)
+		if invoice.Private {
+			totalHopHints = maxHopHints
 		}
 
-		if len(openChannels) > 0 {
-			// We filter the channels by excluding the ones that
-			// were specified by the caller and were already added.
-			var filteredChannels []*HopHintInfo
-			for _, c := range openChannels {
-				scid := c.ShortChanID().ToUint64()
-				if _, ok := forcedHints[scid]; ok {
-					continue
-				}
-
-				// If this is a zero-conf channel, check if the
-				// confirmed SCID was used in forcedHints.
-				if c.IsZeroConf() {
-					scid := c.ZeroConfRealScid().ToUint64()
-					if _, ok := forcedHints[scid]; ok {
-						continue
-					}
-				}
-
-				chanID := lnwire.NewChanIDFromOutPoint(
-					&c.FundingOutpoint,
-				)
-
-				// Check whether the the peer's alias was
-				// provided in forcedHints.
-				peerAlias, _ := cfg.GetAlias(chanID)
```

Comment:
IIUC, this is what causes the slown down. Given we know all the channels here ahead of time, what if we just did a batch request to get all the _potential_ aliases for each of the channels?

Alternatively, we can add a write-thru cache to the Alias Manager, so then `GetPeerAlias` would would just read things in memory. Might be an over all smaller change.

Reply:
Write-thru cache [added](https://github.com/lightningnetwork/lnd/pull/6914/commits/e078f683930e9b0bd6c7c86f86d89a8af7d1452a) :+1: